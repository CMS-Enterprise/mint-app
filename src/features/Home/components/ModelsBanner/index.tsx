import React from 'react';
import { useTranslation } from 'react-i18next';
import { Button, ButtonGroup, Grid } from '@trussworks/react-uswds';
import classNames from 'classnames';
import { helpSolutions } from 'features/HelpAndKnowledge/SolutionsHelp/solutionsMap';
import { ModelsType } from 'features/Home/components/ModelCardTable';
import {
  ComponentGroup,
  GeneralStatus,
  MtoCommonSolutionKey
} from 'gql/generated/graphql';

import useCheckResponsiveScreen from 'hooks/useCheckMobile';

import '../ModelsBySolution/index.scss';

export type StatusCategories =
  | 'total'
  | GeneralStatus.PLANNED
  | GeneralStatus.ACTIVE
  | GeneralStatus.ENDED;

const ModelsBanner = ({
  type,
  filterKey,
  models,
  selectedStatus,
  setSelectedStatus
}: {
  type: 'solution' | 'group';
  filterKey: MtoCommonSolutionKey | ComponentGroup;
  models: ModelsType;
  selectedStatus: StatusCategories;
  setSelectedStatus: (status: StatusCategories) => void;
}) => {
  const { t: customHomeT } = useTranslation('customHome');
  const { t: homepageSettingsT } = useTranslation('homepageSettings');

  const isMobile = useCheckResponsiveScreen('tablet', 'smaller');

  const headerText = () => {
    if (type === 'solution') {
      const selectedSolution = helpSolutions[filterKey as MtoCommonSolutionKey];

      return (
        <>
          {' '}
          {selectedSolution?.name}{' '}
          {selectedSolution?.acronym ? `(${selectedSolution?.acronym})` : ''}
        </>
      );
    }
    return homepageSettingsT(`componentGroupAcronyms.${filterKey}`);
  };

  return (
    <div className="models-by-solutions">
      <div className="bg-primary-lighter radius-md padding-2">
        <Grid row className="flex-align-center">
          <Grid desktop={{ col: 7 }} tablet={{ col: 6 }}>
            <h3 className="margin-y-1">{headerText()}</h3>
          </Grid>

          <Grid
            desktop={{ col: 5 }}
            tablet={{ col: 6 }}
            className={classNames('models-by-solutions__banner-buttons', {
              'row-reverse': !isMobile
            })}
          >
            <ButtonGroup type="segmented">
              <Button
                type="button"
                onClick={() => setSelectedStatus('total')}
                className={classNames('padding-y-2', {
                  'bg-primary-darker': selectedStatus === 'total'
                })}
              >
                <div>{customHomeT(`generalStatus.${GeneralStatus.OTHER}`)}</div>
                <div data-testid="total-count">{models.length}</div>
              </Button>
              <Button
                type="button"
                onClick={() => setSelectedStatus(GeneralStatus.PLANNED)}
                className={classNames('padding-y-2', {
                  'bg-primary-darker': selectedStatus === GeneralStatus.PLANNED
                })}
              >
                <div>
                  {customHomeT(`generalStatus.${GeneralStatus.PLANNED}`)}
                </div>
                <div data-testid="planned-count">
                  {
                    models.filter(
                      model => model.generalStatus === GeneralStatus.PLANNED
                    ).length
                  }
                </div>
              </Button>
              <Button
                type="button"
                onClick={() => setSelectedStatus(GeneralStatus.ACTIVE)}
                className={classNames('padding-y-2', {
                  'bg-primary-darker': selectedStatus === GeneralStatus.ACTIVE
                })}
              >
                <div>
                  {customHomeT(`generalStatus.${GeneralStatus.ACTIVE}`)}
                </div>
                <div data-testid="active-count">
                  {
                    models.filter(
                      model => model.generalStatus === GeneralStatus.ACTIVE
                    ).length
                  }
                </div>
              </Button>
              <Button
                type="button"
                onClick={() => setSelectedStatus(GeneralStatus.ENDED)}
                className={classNames('padding-y-2', {
                  'bg-primary-darker': selectedStatus === GeneralStatus.ENDED
                })}
              >
                <div>{customHomeT(`generalStatus.${GeneralStatus.ENDED}`)}</div>
                <div data-testid="ended-count">
                  {
                    models.filter(
                      model => model.generalStatus === GeneralStatus.ENDED
                    ).length
                  }
                </div>
              </Button>
            </ButtonGroup>
          </Grid>
        </Grid>
      </div>
    </div>
  );
};

export default ModelsBanner;
