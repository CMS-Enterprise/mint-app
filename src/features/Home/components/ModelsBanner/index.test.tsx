import React from 'react';
import { render, screen } from '@testing-library/react';
import {
  ComponentGroup,
  GeneralStatus,
  ModelStatus,
  MtoCommonSolutionKey
} from 'gql/generated/graphql';
import setup from 'tests/util';

import ModelsBanner from './index';

const mockModels = [
  {
    id: 'model-1',
    modelName: 'Model 1',
    status: ModelStatus.PLAN_DRAFT,
    generalStatus: GeneralStatus.PLANNED,
    abbreviation: 'M1',
    basics: {
      id: 'basics-1',
      modelCategory: null,
      __typename: 'PlanBasics' as const
    },
    timeline: {
      id: 'timeline-1',
      performancePeriodStarts: null,
      performancePeriodEnds: null,
      __typename: 'PlanTimeline' as const
    },
    __typename: 'ModelPlan' as const
  },
  {
    id: 'model-2',
    modelName: 'Model 2',
    status: ModelStatus.ACTIVE,
    generalStatus: GeneralStatus.ACTIVE,
    abbreviation: 'M2',
    basics: {
      id: 'basics-2',
      modelCategory: null,
      __typename: 'PlanBasics' as const
    },
    timeline: {
      id: 'timeline-2',
      performancePeriodStarts: '2024-01-01T00:00:00Z',
      performancePeriodEnds: '2025-12-31T00:00:00Z',
      __typename: 'PlanTimeline' as const
    },
    __typename: 'ModelPlan' as const
  },
  {
    id: 'model-3',
    modelName: 'Model 3',
    status: ModelStatus.ENDED,
    generalStatus: GeneralStatus.ENDED,
    abbreviation: 'M3',
    basics: {
      id: 'basics-3',
      modelCategory: null,
      __typename: 'PlanBasics' as const
    },
    timeline: {
      id: 'timeline-3',
      performancePeriodStarts: '2022-01-01T00:00:00Z',
      performancePeriodEnds: '2023-12-31T00:00:00Z',
      __typename: 'PlanTimeline' as const
    },
    __typename: 'ModelPlan' as const
  },
  {
    id: 'model-4',
    modelName: 'Model 4',
    status: ModelStatus.PLAN_COMPLETE,
    generalStatus: GeneralStatus.PLANNED,
    abbreviation: 'M4',
    basics: {
      id: 'basics-4',
      modelCategory: null,
      __typename: 'PlanBasics' as const
    },
    timeline: {
      id: 'timeline-4',
      performancePeriodStarts: null,
      performancePeriodEnds: null,
      __typename: 'PlanTimeline' as const
    },
    __typename: 'ModelPlan' as const
  }
];

describe('ModelsBanner', () => {
  it('renders with solution type and displays solution name', () => {
    const mockSetStatus = vi.fn();

    render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    // Should display solution name and acronym
    expect(screen.getByText(/4innovation/i)).toBeInTheDocument();
  });

  it('renders with group type and displays component group', () => {
    const mockSetStatus = vi.fn();

    render(
      <ModelsBanner
        type="group"
        filterKey={ComponentGroup.CCMI_PCMG}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    // Should display translated component group name
    expect(screen.getByText('CMMI/PCMG')).toBeInTheDocument();
  });

  it('displays correct counts for each status', () => {
    const mockSetStatus = vi.fn();

    render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    // Total: all 4 models
    expect(screen.getByTestId('total-count')).toHaveTextContent('4');

    // Planned: 2 models (model-1 and model-4)
    expect(screen.getByTestId('planned-count')).toHaveTextContent('2');

    // Active: 1 model (model-2)
    expect(screen.getByTestId('active-count')).toHaveTextContent('1');

    // Ended: 1 model (model-3)
    expect(screen.getByTestId('ended-count')).toHaveTextContent('1');
  });

  it('highlights the selected status button', () => {
    const mockSetStatus = vi.fn();

    const { rerender } = render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    // Total button should be highlighted
    const totalButton = screen.getByTestId('total-count').closest('button');
    expect(totalButton).toHaveClass('bg-primary-darker');

    // Rerender with different selected status
    rerender(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus={GeneralStatus.ACTIVE}
        setSelectedStatus={mockSetStatus}
      />
    );

    // Active button should now be highlighted
    const activeButton = screen.getByTestId('active-count').closest('button');
    expect(activeButton).toHaveClass('bg-primary-darker');

    // Total button should no longer be highlighted
    const totalButtonAfter = screen
      .getByTestId('total-count')
      .closest('button');
    expect(totalButtonAfter).not.toHaveClass('bg-primary-darker');
  });

  it('calls setSelectedStatus when status buttons are clicked', async () => {
    const mockSetStatus = vi.fn();

    const { user } = setup(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    // Click on Planned button
    const plannedButton = screen.getByTestId('planned-count').closest('button');
    if (plannedButton) {
      await user.click(plannedButton);
    }
    expect(mockSetStatus).toHaveBeenCalledWith(GeneralStatus.PLANNED);

    // Click on Active button
    const activeButton = screen.getByTestId('active-count').closest('button');
    if (activeButton) {
      await user.click(activeButton);
    }
    expect(mockSetStatus).toHaveBeenCalledWith(GeneralStatus.ACTIVE);

    // Click on Ended button
    const endedButton = screen.getByTestId('ended-count').closest('button');
    if (endedButton) {
      await user.click(endedButton);
    }
    expect(mockSetStatus).toHaveBeenCalledWith(GeneralStatus.ENDED);

    // Click on Total button
    const totalButton = screen.getByTestId('total-count').closest('button');
    if (totalButton) {
      await user.click(totalButton);
    }
    expect(mockSetStatus).toHaveBeenCalledWith('total');

    expect(mockSetStatus).toHaveBeenCalledTimes(4);
  });

  it('displays zero counts when no models match a status', () => {
    const mockSetStatus = vi.fn();

    const plannedOnlyModels = [mockModels[0], mockModels[3]]; // Only planned models

    render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={plannedOnlyModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    expect(screen.getByTestId('total-count')).toHaveTextContent('2');
    expect(screen.getByTestId('planned-count')).toHaveTextContent('2');
    expect(screen.getByTestId('active-count')).toHaveTextContent('0');
    expect(screen.getByTestId('ended-count')).toHaveTextContent('0');
  });

  it('handles empty models array', () => {
    const mockSetStatus = vi.fn();

    render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={[]}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    expect(screen.getByTestId('total-count')).toHaveTextContent('0');
    expect(screen.getByTestId('planned-count')).toHaveTextContent('0');
    expect(screen.getByTestId('active-count')).toHaveTextContent('0');
    expect(screen.getByTestId('ended-count')).toHaveTextContent('0');
  });

  it('renders different component groups correctly', () => {
    const mockSetStatus = vi.fn();

    const groups = [
      ComponentGroup.CCMI_PCMG,
      ComponentGroup.CCMI_PPG,
      ComponentGroup.CCSQ,
      ComponentGroup.CM
    ];

    groups.forEach(group => {
      const { unmount } = render(
        <ModelsBanner
          type="group"
          filterKey={group}
          models={mockModels}
          selectedStatus="total"
          setSelectedStatus={mockSetStatus}
        />
      );

      // Should render the component group heading
      const heading = screen.getByRole('heading', { level: 3 });
      expect(heading).toBeInTheDocument();

      unmount();
    });
  });

  it('renders different solutions correctly', () => {
    const mockSetStatus = vi.fn();

    const solutions = [
      MtoCommonSolutionKey.INNOVATION,
      MtoCommonSolutionKey.ACO_OS,
      MtoCommonSolutionKey.CCW,
      MtoCommonSolutionKey.CBOSC
    ];

    solutions.forEach(solution => {
      const { unmount } = render(
        <ModelsBanner
          type="solution"
          filterKey={solution}
          models={mockModels}
          selectedStatus="total"
          setSelectedStatus={mockSetStatus}
        />
      );

      // Should render the solution heading
      const heading = screen.getByRole('heading', { level: 3 });
      expect(heading).toBeInTheDocument();

      unmount();
    });
  });

  it('matches snapshot with solution type', () => {
    const mockSetStatus = vi.fn();

    const { asFragment } = render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with group type', () => {
    const mockSetStatus = vi.fn();

    const { asFragment } = render(
      <ModelsBanner
        type="group"
        filterKey={ComponentGroup.CCMI_PCMG}
        models={mockModels}
        selectedStatus="total"
        setSelectedStatus={mockSetStatus}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with active status selected', () => {
    const mockSetStatus = vi.fn();

    const { asFragment } = render(
      <ModelsBanner
        type="solution"
        filterKey={MtoCommonSolutionKey.INNOVATION}
        models={mockModels}
        selectedStatus={GeneralStatus.ACTIVE}
        setSelectedStatus={mockSetStatus}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
