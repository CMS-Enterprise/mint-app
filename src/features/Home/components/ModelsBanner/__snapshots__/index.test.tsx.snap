// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelsBanner > matches snapshot with active status selected 1`] = `
<DocumentFragment>
  <div
    class="models-by-solutions"
  >
    <div
      class="bg-primary-lighter radius-md padding-2"
    >
      <div
        class="grid-row flex-align-center"
        data-testid="grid"
      >
        <div
          class="tablet:grid-col-6 desktop:grid-col-7"
          data-testid="grid"
        >
          <h3
            class="margin-y-1"
          >
             4innovation (4i)
          </h3>
        </div>
        <div
          class="tablet:grid-col-6 desktop:grid-col-5 models-by-solutions__banner-buttons row-reverse"
          data-testid="grid"
        >
          <ul
            class="usa-button-group usa-button-group--segmented"
          >
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Total
                </div>
                <div
                  data-testid="total-count"
                >
                  4
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Planned
                </div>
                <div
                  data-testid="planned-count"
                >
                  2
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2 bg-primary-darker"
                data-testid="button"
                type="button"
              >
                <div>
                  Active
                </div>
                <div
                  data-testid="active-count"
                >
                  1
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Ended
                </div>
                <div
                  data-testid="ended-count"
                >
                  1
                </div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelsBanner > matches snapshot with group type 1`] = `
<DocumentFragment>
  <div
    class="models-by-solutions"
  >
    <div
      class="bg-primary-lighter radius-md padding-2"
    >
      <div
        class="grid-row flex-align-center"
        data-testid="grid"
      >
        <div
          class="tablet:grid-col-6 desktop:grid-col-7"
          data-testid="grid"
        >
          <h3
            class="margin-y-1"
          >
            CMMI/PCMG
          </h3>
        </div>
        <div
          class="tablet:grid-col-6 desktop:grid-col-5 models-by-solutions__banner-buttons row-reverse"
          data-testid="grid"
        >
          <ul
            class="usa-button-group usa-button-group--segmented"
          >
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2 bg-primary-darker"
                data-testid="button"
                type="button"
              >
                <div>
                  Total
                </div>
                <div
                  data-testid="total-count"
                >
                  4
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Planned
                </div>
                <div
                  data-testid="planned-count"
                >
                  2
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Active
                </div>
                <div
                  data-testid="active-count"
                >
                  1
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Ended
                </div>
                <div
                  data-testid="ended-count"
                >
                  1
                </div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelsBanner > matches snapshot with solution type 1`] = `
<DocumentFragment>
  <div
    class="models-by-solutions"
  >
    <div
      class="bg-primary-lighter radius-md padding-2"
    >
      <div
        class="grid-row flex-align-center"
        data-testid="grid"
      >
        <div
          class="tablet:grid-col-6 desktop:grid-col-7"
          data-testid="grid"
        >
          <h3
            class="margin-y-1"
          >
             4innovation (4i)
          </h3>
        </div>
        <div
          class="tablet:grid-col-6 desktop:grid-col-5 models-by-solutions__banner-buttons row-reverse"
          data-testid="grid"
        >
          <ul
            class="usa-button-group usa-button-group--segmented"
          >
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2 bg-primary-darker"
                data-testid="button"
                type="button"
              >
                <div>
                  Total
                </div>
                <div
                  data-testid="total-count"
                >
                  4
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Planned
                </div>
                <div
                  data-testid="planned-count"
                >
                  2
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Active
                </div>
                <div
                  data-testid="active-count"
                >
                  1
                </div>
              </button>
            </li>
            <li
              class="usa-button-group__item"
            >
              <button
                class="usa-button padding-y-2"
                data-testid="button"
                type="button"
              >
                <div>
                  Ended
                </div>
                <div
                  data-testid="ended-count"
                >
                  1
                </div>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
